agent/util-scripts/pbench-tool-data-sink

```diff
@@ -0,0 +1,551 @@
+                )
+            remaining_bytes = content_length
+
+        try:
+            exp_md5 = request["HTTP_MD5SUM"]
+        except Exception:
+            self.logger.exception(request.keys())
+            abort(400, "Missing required md5sum header")
+
+        target_dir = self.benchmark_run_dir / self.directory
+        if not target_dir.is_dir():
+            self.logger.error("ERROR - directory, '%s', does not exist", target_dir)
+            abort(500, f"Invalid URL, path {target_dir} does not exist")
+        host_tool_data_tb_name = target_dir / f"{hostname}.tar.xz"
+        if host_tool_data_tb_name.exists():
+            abort(409, f"{host_tool_data_tb_name} already uploaded")
+        host_tool_data_tb_md5 = Path(f"{host_tool_data_tb_name}.md5")
+
+        with tempfile.NamedTemporaryFile(mode="wb", dir=target_dir) as ofp:
+            total_bytes = 0
+            iostr = request["wsgi.input"]
+            h = hashlib.md5()
+            while remaining_bytes > 0:
+                buf = iostr.read(
+                    _BUFFER_SIZE if remaining_bytes > _BUFFER_SIZE else remaining_bytes
+                )
+                bytes_read = len(buf)
+                total_bytes += bytes_read
+                remaining_bytes -= bytes_read
+                h.update(buf)
+                ofp.write(buf)
+            cur_md5 = h.hexdigest()
+            if cur_md5 != exp_md5:
+                abort(
+                    400,
+                    f"Content, {cur_md5}, does not match its MD5SUM header,"
+                    f" {exp_md5}",
+                )
+            if total_bytes <= 0:
+                abort(400, "No data received")
+
+            # First write the .md5
+            try:
+                with host_tool_data_tb_md5.open("w") as md5fp:
+                    md5fp.write(f"{exp_md5} {host_tool_data_tb_name.name}\n")
+            except Exception:
+                try:
+                    os.remove(host_tool_data_tb_md5)
+                except Exception:
+                    self.logger.warning(
+                        "Failed to remove .md5 %s when trying to clean" " up",
+                        host_tool_data_tb_md5,
+                    )
+                raise
+
+            # Then create the final filename link to the temporary file.
+            os.link(ofp.name, host_tool_data_tb_name)
+            self.logger.info(
+                "Successfully wrote %s (%s.md5)",
+                host_tool_data_tb_name,
+                host_tool_data_tb_name,
+            )
+            # Tell the waiting "watcher" thread engaging in a "state" change
+            # that another PUT document has arrived.
+            with self._lock:
+                tm_tracker = self._tm_tracking[hostname]
+                assert tm_tracker["posted"] == "waiting", f"tm_tracker = {tm_tracker!r}"
+                tm_tracker["posted"] = "yes"
+                self._put_count += 1
+                self._cv.notify()
+
+
+def main(argv):
```

Comment:
This bit should be in the /lib/pbench/cli/agent

Reply:
Agreed.  Avoided more change...